cmd/registry-replacer/main.go

```diff
@@ -0,0 +1,210 @@
+		if !referencesDockerfile(config.Images) {
+			return nil
+		}
+
+		originalConfig, err := yaml.Marshal(config)
+		if err != nil {
+			return fmt.Errorf("failed to marshal config for comparison: %w", err)
+		}
+
+		for idx := range config.Images {
+			foundTags, err := ensureReplacement(&config.Images[idx], githubFileGetterFactory(info.Org, info.Repo, info.Branch))
+			if err != nil {
+				return fmt.Errorf("failed to ensure replacements: %w", err)
+			}
+			for _, foundTag := range foundTags {
+				if config.BaseImages == nil {
+					config.BaseImages = map[string]api.ImageStreamTagReference{}
+				}
+				config.BaseImages[foundTag.String()] = api.ImageStreamTagReference{
+					Namespace: foundTag.org,
+					Name:      foundTag.repo,
+					Tag:       foundTag.tag,
+				}
+			}
+		}
+
+		newConfig, err := yaml.Marshal(config)
+		if err != nil {
+			return fmt.Errorf("failed to marshal new config: %w", err)
+		}
+
+		// Avoid filesystem access if possible
+		if bytes.Equal(originalConfig, newConfig) {
+			return nil
+		}
+
+		if err := writer(info.Filename, newConfig, 0644); err != nil {
+			return fmt.Errorf("faild to write %s: %w", info.Filename, err)
+		}
+
+		return nil
+	}
+}
+
+func referencesDockerfile(images []api.ProjectDirectoryImageBuildStepConfiguration) bool {
+	for _, image := range images {
+		if image.DockerfilePath != "" {
+			return true
+		}
+	}
+
+	return false
+}
+
+var registryRegex = regexp.MustCompile(`registry\.svc\.ci\.openshift\.org\/[^\s]+`)
+
+type orgRepoTag struct{ org, repo, tag string }
+
+func (ort orgRepoTag) String() string {
+	return ort.org + "_" + ort.repo + "_" + ort.tag
+}
+
+func ensureReplacement(image *api.ProjectDirectoryImageBuildStepConfiguration, getter githubFileGetter) ([]orgRepoTag, error) {
+	if image.DockerfilePath == "" {
+		return nil, nil
+	}
+
+	data, err := getter(image.DockerfilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get dockerfile %s: %w", image.DockerfilePath, err)
+	}
+
+	var toReplace []string
+	for _, line := range bytes.Split(data, []byte("\n")) {
+		if !bytes.Contains(line, []byte("FROM")) {
+			continue
+		}
+		if !bytes.Contains(line, []byte("registry.svc.ci.openshift.org")) {
+			continue
+		}
+		match := registryRegex.Find(line)
+		if match == nil {
+			continue
+		}
+
+		toReplace = append(toReplace, string(match))
+	}
+
+	var result []orgRepoTag
+	for _, toReplace := range toReplace {
+		orgRepoTag, err := orgRepoTagFromPullString(toReplace)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse string %s as pullspec: %w", toReplace, err)
+		}
+		if image.Inputs == nil {
+			image.Inputs = map[string]api.ImageBuildInputs{}
+		}
+		inputs := image.Inputs[orgRepoTag.String()]
+		inputs.As = sets.NewString(inputs.As...).Insert(toReplace).List()
+		image.Inputs[orgRepoTag.String()] = inputs
+
+		result = append(result, orgRepoTag)
+	}
+
+	return result, nil
+}
+
+func orgRepoTagFromPullString(pullString string) (orgRepoTag, error) {
```

Comment:
I imagine there's a canonical utility for this somewhere but it's simple enough so if we can't find one quickly let's keep our own impl